### PR TITLE
Adding Analytics commissions

### PR DESCRIPTION
**Pull request by IPadawans**
#### Analytics Commissions Implementation

1.  **Distribution Logic**
    
    -   Commissions for the `AnalyticsSubscriptionBounty` will now be distributed after a new billingId from API.
2.  **Device Selection**
    
    -   Commissions for a team will be directed to a single device, but we need to calculate the average balance of all devices of the team and use this average as collateral.

#### Payment Conditions

1.  **Frequency**
    
    -   The `AnalyticsSubscriptionBounty` will be paid after a new billingId.
2.  **BillingID Criteria**
    
    -   Payment will only proceed if the billingId of the device does not exist or is different from the last billingId after the initial commission distribution.

#### Next Check-in Logic After Distributing First Commission

1.  **Check for BillingID**
    
    -   On the next check-in, validate if the billingId is different from the previous one.
2.  **Update `nextEpochProgressToRewardAnalytics`**
    
    -   If the billingId is different, set `nextEpochProgressToRewardAnalytics` to be distributed on the next daily distribution with an offset of 2.

#### Offset Explanation

-   0: InstallationBounty
-   1: CommercialBounty
-   2: AnalyticsBounty

### Summary

These changes enhance the commission structure by introducing Analytics commissions for the `AnalyticsSubscriptionBounty`. The distribution logic ensures that commissions are directed to a single device within a team, but using the collateral as an average of balances of all devices of the team. Payment conditions specify that the `AnalyticsSubscriptionBounty` will be paid after a new billingId. The next check-in logic updates the `nextEpochProgressToRewardAnalytics` based on the billingId difference, with an offset indicating the type of commission.

### CODE REVIEW

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/DailyBountyRewards.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import com.my.dor_metagraph.shared_data.Utils._
+import com.my.dor_metagraph.shared_data.bounties.{CommercialLocationBounty, UnitDeployedBounty}
+import com.my.dor_metagraph.shared_data.types.Types._
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.transaction.RewardTransaction
+import org.typelevel.log4cats.SelfAwareStructuredLogger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+case class DailyBountyRewards() extends BountyRewards {
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/DailyBountyRewards.scala]: Thanks for the tip, updated to follow the pattern

File: metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/combiners/DeviceCheckIn.scala [hunk 2]
```diff
@@ -35,4 +31,63 @@ object DeviceCheckIn {
...
+    maybeDeviceInfo: Option[DeviceInfo],
+    dorAPIResponse : DorAPIResponse
+  ): Option[AnalyticsBountyInformation] =
+    dorAPIResponse.lastBillingId.map { lastBillingId =>
+      maybeDeviceInfo.fold(createAnalyticsBountyInformation(epochProgress, dorAPIResponse)) { deviceInfo =>
+        deviceInfo.analyticsBountyInformation.fold(createAnalyticsBountyInformation(epochProgress, dorAPIResponse)) { oldAnalyticsBountyInformation =>
+          updateAnalyticsBountyInformation(epochProgress, dorAPIResponse, lastBillingId, oldAnalyticsBountyInformation)
+        }
+      }
+    }
```
Comment (sdavidp) [inline metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/combiners/DeviceCheckIn.scala:65]: This is a bit hard to read. Is this equivalent?
```suggestion
    dorAPIResponse
      .lastBillingId
      .map { lastBillingId =>
        maybeDeviceInfo
          .flatMap(_.analyticsBountyInformation)
          .map(updateAnalyticsBountyInformation(epochProgress, dorAPIResponse, lastBillingId, _))
          .getOrElse(createAnalyticsBountyInformation(epochProgress, dorAPIResponse))
      }
```

File: metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/combiners/DeviceCheckIn.scala
Comment (sdavidp) [inline metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/combiners/DeviceCheckIn.scala:1]: This class has many calls to `Option.get` which should only be used rarely.
It's not clear by looking at the code in this module if those options are never `None` so the module likely needs to be refactored.

**Review by sdavidp — CHANGES_REQUESTED** [review]

**Comment by sdavidp** [comment]
Also, look for places where there's a `>> something.pure[F]` as that can be replaced by `.as(something)`

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/DorRewards.scala [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import org.tessellation.security.signature.Signed
+import org.typelevel.log4cats.SelfAwareStructuredLogger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+import scala.collection.immutable.{SortedMap, SortedSet}
+
+object DorRewards {
+  def logger[F[_] : Async]: SelfAwareStructuredLogger[F] = Slf4jLogger.getLoggerFromName[F]("MainRewards")
+
+  def make[F[_] : Async : SecurityProvider]: Rewards[F, CurrencySnapshotStateProof, CurrencyIncrementalSnapshot, CurrencySnapshotEvent] = {
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/DorRewards.scala:30]: Added

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/DorRewards.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import org.tessellation.sdk.infrastructure.consensus.trigger.ConsensusTrigger
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signed
+import org.typelevel.log4cats.SelfAwareStructuredLogger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+import scala.collection.immutable.{SortedMap, SortedSet}
+
+object DorRewards {
+  def logger[F[_] : Async]: SelfAwareStructuredLogger[F] = Slf4jLogger.getLoggerFromName[F]("MainRewards")
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/DorRewards.scala:28]: Fixed

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/DorRewards.scala
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/DorRewards.scala:1]: Yeah, you're probably right. I've refactored and adapted the tests

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          }
+
+        for {
+          _ <- logInitialRewardDistribution(currentEpochProgress)
+
+          groupedDevices = state.devices.collect {
+              case (_, info) if info.analyticsBountyInformation.exists(_.nextEpochProgressToRewardAnalytics == currentEpochProgress) => info
+            }
+            .groupBy(_.analyticsBountyInformation.map(_.teamId).getOrElse(UndefinedTeamId))
+            .filter(_._1 != UndefinedTeamId)
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala:59]: Yeah, fixed

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+          analyticsRewardTransactionsAndValidatorTaxes <- groupedDevices
+            .toList
+            .foldLeftM(RewardTransactionsInformation(Map.empty, 0L, lastBalancesRaw)) { (acc, devices) =>
+              val teamDevices = devices._2.toList
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala:64]: Fixed

File: metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/bounties/AnalyticsSubscriptionBounty.scala [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.my.dor_metagraph.shared_data.bounties
+
+import com.my.dor_metagraph.shared_data.types.Types.DorAPIResponse
+
+case class AnalyticsSubscriptionBounty() extends Bounty {
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/bounties/AnalyticsSubscriptionBounty.scala:5]: Updated to class

File: metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/bounties/AnalyticsSubscriptionBounty.scala [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  override def getBountyRewardAmount(
+    deviceInfo  : DorAPIResponse,
+    epochModulus: Long
+  ): Double = {
+    deviceInfo.billedAmount match {
+      case None => 0D
+      case Some(value) =>
+        value * 25D
+    }
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/bounties/AnalyticsSubscriptionBounty.scala:14]: Fixed

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            toTokenAmountFormat(AnalyticsSubscriptionBounty().getBountyRewardAmount(device.dorAPIResponse, 0L))
+          } else {
+            0L
+          }
+        }
+      }
+
+      override def logInitialRewardDistribution(
+        epochProgressModulus: Long
+      ): F[Unit] = logger.info("Starting AnalyticsSubscriptionBounty bounty distribution")
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala:107]: Updated

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/BountyRewards.scala [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.transaction.RewardTransaction
+
+abstract class BountyRewards[F[_]: Async] {
+
+  val ValidatorNodeTaxRate: Double = 0.10
+
+  def logInitialRewardDistribution(
+    currentEpochProgress: Long
+  ): F[Unit] = ().pure[F]
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/BountyRewards.scala:19]: Updated

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    val devicesBalances = devices.map { device =>
+      device.dorAPIResponse.rewardAddress.fold(0L) { rewardAddress =>
+        balances.get(rewardAddress)
+          .map(_.value.value)
+          .getOrElse(0L)
+      }
+    }.sum
+
+    devicesBalances / devices.size
+  }
```
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/bounties/AnalyticsBountyRewards.scala:140]: Yeah, updated

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/validators/ValidatorNodes.scala
[the remark this answers is not in the document]
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/validators/ValidatorNodes.scala:1]: I've refactored to be a service injected on the rewards too, to become easy to test, but I've applied your suggestion too

File: metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/combiners/DeviceCheckIn.scala (hunk 2 above)
Comment (IPadawans) [reply metagraph/modules/shared_data/src/main/scala/com/my/dor_metagraph/shared_data/combiners/DeviceCheckIn.scala:65]: Yeah, thanks

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/Main.scala [hunk 12]
```diff
@@ -145,7 +148,19 @@ object Main extends CurrencyL0App(
   override def dataApplication: Option[Resource[IO, BaseDataApplicationL0Service[IO]]] =
     makeL0Service.asResource.some
 
-  override def rewards(implicit sp: SecurityProvider[IO]): Option[Rewards[IO, CurrencySnapshotStateProof, CurrencyIncrementalSnapshot, CurrencySnapshotEvent]] =
-    DorRewards.make[IO].some
+  override def rewards(implicit sp: SecurityProvider[IO]): Option[Rewards[IO, CurrencySnapshotStateProof, CurrencyIncrementalSnapshot, CurrencySnapshotEvent]] = {
+    val dailyBountyRewards = new DailyBountyRewards[IO]
+    val analyticsBountyRewards = new AnalyticsBountyRewards[IO]
+    val l0Url = getEnv[IO]("METAGRAPH_L0_NODE_URL")
+    val l1Url = getEnv[IO]("DATA_L1_NODE_URL")
```
Comment (sdavidp) [inline metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/Main.scala:155]: This doesn't change the effective functionality -- you're still executing the getEnv on every call. These should be higher up.

I see the confusion though -- we aren't defining `Main` like most apps do, so you need to put this in the `makeL0Service` so you have access to the results, then pass that to whoever needs it. E.g.
```
  private def makeL0Service: IO[BaseDataApplicationL0Service[IO]] = {
    for {
      l0Url <- getEnv[IO]("METAGRAPH_L0_NODE_URL")
      l1Url <- getEnv[IO]("DATA_L1_NODE_URL")
      validatorApi = new ValidatorNodesAPI[IO](l0Url, l1Url)
      //  then you can pass the api where needed
      calculatedStateService <- CalculatedStateService.make[IO]
      dataApplicationL0Service = makeBaseDataApplicationL0Service(calculatedStateService)
    } yield dataApplicationL0Service
  }
```

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/validators/ValidatorNodesAPI.scala [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+import cats.effect.Async
+import cats.syntax.all._
+import com.my.dor_metagraph.shared_data.external_apis.ClusterApi.getValidatorNodesAddresses
+import org.tessellation.schema.address.Address
+import org.tessellation.security.SecurityProvider
+
+class ValidatorNodesAPI[F[_] : Async : SecurityProvider](
+  l0UrlF: F[String],
+  l1UrlF: F[String]
```
Comment (sdavidp) [inline metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/validators/ValidatorNodesAPI.scala:11]:
```suggestion
  l0Url: String,
  l1Url: String
```

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/validators/ValidatorNodesAPI.scala [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+) extends ValidatorNodes[F] {
+
+  override def getValidatorNodes: F[(List[Address], List[Address])] =
+    for {
+      l0Url <- l0UrlF
+      l1Url <- l1UrlF
+      validatorsAddresses <- getValidatorNodesAddresses(l0Url, l1Url)
+    } yield validatorsAddresses
```
Comment (sdavidp) [inline metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/rewards/validators/ValidatorNodesAPI.scala:19]:
```suggestion
getValidatorNodesAddresses(l0Url, l1Url)
```

**Review by sdavidp — CHANGES_REQUESTED** [review]
Couple small changes otherwise looks good

File: metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/Main.scala (hunk 12 above)
Comment (IPadawans) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/Main.scala:155]: @sdavidp the rewards function does not belong to this function, I mean it's not from `BaseDataApplicationL0Service`

Comment (sdavidp) [reply metagraph/modules/l0/src/main/scala/com/my/dor_metagraph/l0/Main.scala]: @IPadawans I see now. Ok, as discussed we're stuck having to do that on each invocation.

**Review by sdavidp — APPROVED** [review]

**Pull request closed without merging**